### DISCUSSION TITLE
How to reload an editor widget

### DISCUSSION THREAD

**Discussion by MikaelDesharnais-ST** [eclipse-theia/theia#8520] [General] [answered]
Hi,

I have the following code :
`
let file = new URI("mem-txt://....");  
let widget = await this.editorManager.open(file);
`
And I would like to reload the editor at some point with some new data (a new URI for example) without having to close the current editor and open a new one.
How should I do ?

I've looked at how the FileResource manages this on file change but I can't understand how this all works.

Regards,

**Comment by vince-fugnitto** [comment]
@MikaelDesharnais-ST I'm not sure I understand this comment:

> And I would like to reload the editor at some point with some new data (a new URI for example) without having to close the current editor and open a new one.

I'm not sure what you mean by reloading the editor, editors correspond to a single file resource. If you require opening a new editor without having to close the old one then you would call the `editorManager.open` again for the new resource.

**Comment by MikaelDesharnais-ST** [comment]
To take another example, when you open a file in an editor, it gets updated when the file is changed.

What I'm trying to achieve is the same thing but for some data stored in memory : Let's say I have a registry that stores some content to be displayed in an editor. If the content changes I would like to update the editor's content.

**Reply by vince-fugnitto** [reply]
Based on your description it looks like you will need to implement a custom editor which uses in-memory data rather than on the filesystem.
You can take a look at some past discussions regarding custom editors such as: https://community.theia-ide.org/t/custom-editor-workflow-examples/290/2.

Your custom editor would need to watch or listen to changes to it's source in order to update itself.

**Comment by MikaelDesharnais-ST** [comment, marked as answer]
I've found the solution to my problem :

I've created a new type of resource resolver for the following scheme : bmem-txt://
This resource is able to get data from a specific registry that acts like a basic hashmap<string,string> and fires event if a key is modified.

The resource extends MutableResource and fires fireDidChangeContents if a change is made in the registry. The implementation of the readContents method for this resource simply returns the content of the registry for the key corresponding to the URI map.

Thus, for a specific URI, only one editor can be opened and if the content corresponding the URI is changed in the registry, the editor is reloaded. Even better, the scroll position seems to be kept (Great) !

The code that I finally found and that solved it all was in : packages\monaco\src\browser\monaco-editor-model.ts
            if (this.resource.onDidChangeContents) {
                this.toDispose.push(this.resource.onDidChangeContents(() => this.sync()));
            }
